### ISSUE TITLE
[MIR / Codegen] Lower a counted loop with its trip-count exit at the latch, not the header

### ISSUE THREAD

**Issue by sanohiro** [sanohiro/align#1084]
Baselines: Align `8c8bfbc7a3169e84ecc8415f5149ab8c61afe863` (alignc 0.7.5, LLVM 22.1.8), align-llm
`b72b0b48`, Apple M1, `--profile release`, default `--target-cpu baseline`. Language question first;
align-llm figures are evidence only.

## Problem

Align has exactly one sequential control form — `loop` with a value-carrying `break`; `for`, `while`,
`continue` and labels are a locked decision against and are not requested here. So the idiomatic
counted loop is —

```align
mut i := 0
loop {
  if i >= n { break }
  ...
  i = i + 1
}
```

and it lowers with the **trip-count test in the loop header** and the increment/back-edge at the bottom. A loop
whose body also exits early (a search, a scan, an `any`/`all`) therefore reaches LLVM as a loop with *two*
header-reachable exits, which LLVM 22's vectorizer structurally rejects:

```
loop not vectorized: Cannot vectorize early exit loop
```

The canonical shape LLVM wants is the rotated one: the header holds only the *data-dependent* test, and the
trip-count test sits at the latch. That is a pure lowering choice — the source is already a counted loop,
and Align throws the information away.

A second, independent fact blocks the rotated form: nothing tells LLVM that a slice's data buffer is
dereferenceable for `len` bytes, so a speculative widened load is illegal
(*"Auto-vectorization of loops with potentially faulting load is not supported"*). Both must be fixed for
the shape to vectorize; each alone gets zero vector instructions.

### 1. The MIR puts the trip-count exit in the header

`kv_plane$all_zero` — an ordinary "are all these bytes zero" scan — from `alignc emit-mir`
(`mir/kv_plane.mir:14112`):

```text
fn all_zero(Borrow _0: slice) -> bool borrow=None region=None cleanup=None {
  bb0:
    _0 <- arg0
    _1 <- 0_i64
    goto bb1
  bb1:
    %0 = load _1
    %1 = load _0
    %2 = slice_len(%1)
    %3 = %0 >= %2
    branch %3 ? bb3 : bb4        <-- TRIP-COUNT EXIT IN THE HEADER
  bb2:
    return true
  bb3:
    goto bb2
  bb4:
    goto bb5
  bb5:
    %4 = load _0
    %5 = load _1
    %6 = slice_len(%4)
    %7 = %5 + 1_i64
    %8 = %5 < 0_i64
    %9 = %5 > %7
    %10 = %7 > %6
    %11 = %8 || %9
    %12 = %11 || %10
    branch %12 ? bb6 : bb7
  bb7:
    %14 = bytes_read_le(%4, %5)
    %15 = %14 as i64 (from u8)
    %16 = %15 != 0_i64
    branch %16 ? bb8 : bb9       <-- DATA EXIT, at the bottom
```

The resulting machine code in the shipped 5.3 MB image is **6 instructions and 2 branches per byte**:

```
000000010004bf84 <_align_fn$17$6b765f706c616e6524616c6c5f7a65726f>:
10004bf84:     	ldp	x9, x8, [x0]
10004bf88:     	mov	x12, xzr
10004bf8c:     	bic	x10, x8, x8, asr #63
<L0>:
10004bf90:     	mov	x11, x12
10004bf94:     	cmp	x10, x12
10004bf98:     	b.eq	 <L1>
10004bf9c:     	ldrb	w13, [x9, x11]
10004bfa0:     	add	x12, x11, #0x1
10004bfa4:     	cbz	w13,  <L0>
<L1>:
10004bfa8:     	cmp	x11, x8
10004bfac:     	cset	w0, ge
10004bfb0:     	ret
```

### 2. Three-way factorization: both factors are required

Taking the *exact* emitted IR for that loop and re-running the identical `default<O2>` pipeline with
`llvm@22/bin/opt`, changing one thing at a time:

| variant | LLVM's own verdict | `vector.body` | width |
|---|---|---|---|
| exact emitted IR (header holds the trip count) | `loop not vectorized: Cannot vectorize early exit loop` | 0 | — |
| rotated: header holds only `p[i] != 0`, latch holds `i+1 == n` | `loop not vectorized: Auto-vectorization of loops with potentially faulting load is not supported` | 0 | — |
| rotated **+** a dereferenceable fact for the data buffer | (no rejection) | **1** | **`<16 x i8>`** |

The dereferenceable fact used in the third variant, verbatim:

```llvm
  call void @llvm.assume(i1 true) ["dereferenceable"(ptr %p, i64 %smax)]
```

Measured on 1 MiB × 2,000 iterations: **3.12 GB/s → 32.7 GB/s (10.5×)**.

Passing `-enable-early-exit-vectorization=true` explicitly changes nothing for the first variant — it is
already reachable in LLVM 22.1.8 and the rejection is structural, not flag-gated. A hand-written C twin of
the same scan hits only the *second* rejection, never the first, because clang emits the rotated shape.

### 3. Scale, as evidence

`alignc explain-opt` over align-llm's hot modules reports **191** `Cannot vectorize early exit loop`
remarks at **140 distinct source sites** across **18 source files** (top: `layer_forward.align` 58,
`moe_layer_forward.align` 43, `model_forward.align` 36, `alignpack_read.align` 10,
`layer_olmoe.align` 9, `runtime_bundle.align` 7), plus **17** `potentially faulting load` remarks at 6
sites (`layer_olmoe.align:1404:5`, `runtime_bundle.align:91:5`, `runtime_execution.align:54:5`,
`model_forward.align:1153:5`, `:2978:5`, `runtime_sampler.align:116:5`). Whole program: **73**
`vector.body` in 2,108 defined Align functions.

Every `any`/`all`/`find`/`first-mismatch`/digest-compare/UTF-8-scan/token-lookup written in the idiomatic
counted-loop form is in this class. This is not a niche shape; it is the second most common loop in any
language.

## Root cause

1. **No canonical counted-loop form.** MIR lowers `loop { if cond { break } … }` literally: the `if` becomes
   a header branch. Nothing recognizes "this is a counted loop with a positive constant step and a
   loop-invariant bound" and rotates it, so LLVM receives a multi-exit loop where a single-exit counted
   loop with one early exit was written.
2. **No dereferenceability fact for slice data.** A `slice<T>` is `{ptr, len}` and the buffer is, by
   Align's own construction, `len * sizeof(T)` accessible bytes. Nothing communicates that, so a widened
   speculative load is illegal even after rotation. A `dereferenceable(16)` attribute on the *header*
   parameter does not help — it describes the 16-byte header, not the buffer the header points at
   (verified).

## Proposed solution

Two changes, stated as language/lowering contracts rather than as a peephole.

**(1) One canonical counted-loop form.** Recognize the counted shape in MIR — a loop whose first statement
is a conditional `break` on `i REL bound` with `i` a slot stepped by a positive constant in the latch and
`bound` loop-invariant — and lower it rotated: peel the zero-trip test into the preheader, keep only the
body's data-dependent exits in the header, and put the `i+step REL bound` back-edge test at the latch.
Nothing about the source language changes and no new syntax is required.

No new surface is proposed, and none is available: *"Sequential control uses one `loop` expression
with value-carrying `break`. There is no `for`, `while`, `continue`, labels, or guaranteed TCO"* is a
locked decision, so a range loop is not the fix and is not asked for. The consequence is that the one
`loop` form must carry the canonical lowering itself: **recognition has to be part of MIR's contract
for `loop`, not a peephole**. Concretely, the recognized counted shape should produce a named MIR
fact — trip count, step, monotone index — that the lowering, the `!range` metadata of #1080 and the
monotone-induction proof of #1081 all read from one place, so "did you write the loop the fast way?"
cannot arise for the only loop the language has. Recognition must be total for the canonical shape
(one spelling in, one shape out) rather than best-effort, or it reintroduces the cliff it removes.

**(2) Emit the dereferenceability fact for slice/array data.** A `slice<T>`'s buffer is `len * sizeof(T)`
dereferenceable bytes by Align's own representation — this is a statement about the compiler's data layout,
not a claim minted from source spelling, so it sits outside the `docs/impl/63-codegen-performance-audit.md:218`
/ `64-composed-byte-optimization-plan.md:186` prohibition on facts derived from syntax. It is also not the
general `llvm.assume` policy rejected at `docs/open-questions.md:4187-4189`: that entry rejects
*"`llvm.assume` / early intrinsic emission / loop-metadata overrides **as a general policy**"* with the
guidance *"attributes and flags first"*. Preferred order here is exactly that:

- **first choice, an attribute**: `dereferenceable` on the data pointer once it is materialized — which
  becomes available for free if #1079 hoists the header load into the preheader, because there is
  then a single `%ptr` SSA value to attribute;
- **fallback, one operand-bundle assume in the preheader**, as in variant 3 above, emitted only for a
  recognized counted loop over a slice and describing only Align's own `{ptr,len}` invariant.

Note the negative result so it is not retried: `dereferenceable(16) align 8` on the `borrow` *header*
parameter does not reach the data buffer.

## Acceptance criteria

Stated in language terms, verified on real kernels:

- The canonical counted loop `loop { if i >= n { break } … i = i + 1 }` lowers with the trip-count
  test at the latch: MIR shows the header containing only body-derived
  branches, and a counted loop with no early exit is unchanged.
- An `all`-style scan over a `slice<u8>` (`kv_plane$all_zero`'s shape) emits a `<16 x i8>` `vector.body`
  at `--profile release --target-cpu baseline`, and measures ≥ 25 GB/s on a 1 MiB buffer (today
  3.12 GB/s).
- `alignc explain-opt` no longer reports `Cannot vectorize early exit loop` for the canonical form; the
  191 remarks at 140 sites in align-llm drop by at least an order of magnitude.
- A `find`-style loop returning the first matching index still returns the *first* match, and a loop whose
  early exit has a side effect (a trap, an `Err`) still takes it at the same iteration. Owner tests for:
  zero-trip, one-trip, early exit on the first element, early exit on the last element, no early exit.
- The dereferenceable fact is emitted only for a recognized slice/array data pointer; a test asserts it is
  absent for a raw FFI pointer and for a zero-length slice.
- No change to any trap message or to `crates/align_driver/tests/vectorize_shapes.rs`.

## Relation to existing issues/plans

- Complementary to, and independent of, the **loop-facts contract** issues — *"[Codegen] Materialize
  a borrowed slice/array view header once and state its alias facts"* (#1079), *"[Codegen] Attach
  `!range` to slice/array length loads so a length is known non-negative"* (#1080), *"[MIR] Fuse
  the bounds check into one unsigned compare and eliminate monotone-induction in-loop checks"*
  (#1081). Those three decide what a loop knows about its *data*; this one decides the loop's
  *control-flow shape*. The dereferenceability half of this issue becomes much cheaper once #1079
  hoists the header load, so scheduling them together is sensible.
- Not covered by `docs/impl/63-codegen-performance-audit.md` (byte-range proof) or
  `64-composed-byte-optimization-plan.md`: both are about eliminating checks, not about loop rotation, and
  neither mentions early-exit vectorization or dereferenceability.
- No overlap with #1047, #1063, #1064, #1065, #1066.
- `docs/open-questions.md:3182-3186` deliberately never asserts `mustprogress`/`willreturn`; nothing here
  requires either.

## Impact framing

Evidence-only scope for align-llm today: Align-compiled code is ~0.1 % of CPU in a real Metal
generation profile, so this is not a tokens/s headline. Where it shows up is the tokenizer path
(**74.5 % Align code**, full of scans), the Align-owned per-token CPU kernels (greedy logits
selection 242 µs/token over 151,936 logits, sampler 267 µs/token over 152,064, this host), the
KV-plane/digest verification scans, and the future in-language CPU inference path. As a *language*
matter the case is stronger and independent of any client: the second most common loop shape in the
language has a lowering that LLVM structurally cannot vectorize, and 10.5× is available on a byte
scan with no change to what the program means.

**Comment by sanohiro** [comment]
align-llm consumer residual at managed Align d9b0df32 (macOS arm64, release, no rt-LTO): the named kv_plane.all_zero witness remains scalar in the exact product image. Its body is 12 arm64 instructions and has the trip-count test in the header: bic clamp, cmp/b.eq, byte load, increment, cbz back to the header; there is no vector body. A function-scoped optimized-IR check after replacing only view.u8(at) with direct view[at] still produces a scalar loop with llvm.smax.i64 and no vector.body, so the source spelling does not consume the promised G3 result. The adjacent #1080 residual (the borrowed-slice length load lacks !range) may be causal, but Plan 69 names this exact client witness and the G3 owner promises the vector result. tokenizer and alignpack smokes pass; no Linux measurement was run.

**Comment by sanohiro** [comment]
Align-side residual is complete in PR #1168 (merge `8e0e5aae32571abb6f061f40840042d904608a0a`).

The exact `kv_plane.all_zero` source shape now authenticates its width-one `BytesRead` guard, preserves the borrowed length `!range` fact, reaches the counted-loop rotation and emits a `<16 x i8>` vector body on both AArch64 baseline and x86-64-v2 owners. Live guards and Plan 64-eliminated guards use separate fail-closed proof paths; wider or mismatched records remain refused. Zero/one/first/last/no-match execution, trap retention, whole/per-unit composition, the bounded gate, Clippy, and Linux x86_64, Linux ARM64, and macOS Apple Silicon CI all passed.

The original align-llm whole-image remark census and 1 MiB throughput measurement remain consumer-owned verification after repinning; they do not leave further Align implementation work on this issue.
